### ISSUE TITLE
Signing Certificate - probable issue

### ISSUE THREAD

**Issue by glawrence** [sqlitebrowser/sqlitebrowser#1229]
I have just read [Termination of the certificates business of Startcom - Google Groups](https://groups.google.com/forum/#!topic/mozilla.dev.security.policy/LM1SpKHJ-oc) and I think we sign with a StartCom certificate, so maybe we need to look at other options?

**Comment by justinclift** [comment]
Thanks @glawrence.  Yep, that will affect us.

The StartCom Certificate we use for signing is valid through to 2018-08-24, so we should be good until around then.  We'll want to have our current release at the time be signed by some other cert provider, so we'll probably need to arrange something by June or July next year.

**Comment by NightmareJoker2** [comment]
The next best option is probably going to be [Certum](https://www.certum.eu/certum/cert,offer_en_open_source_cs.xml) for you, they have an *open source* "deal" at EUR 29/year.

**Comment by justinclift** [comment]
Thanks @NightmareJoker2.  We looked at Certum last time around too, but we need a certificate that is Microsoft Authenticode compliant.  That page has no mention of Authenticode anywhere. 😦 

If it did, we probably would have chosen them last time instead of going with StartCom. :smile:

**Comment by justinclift** [comment]
As a data point, Centum's page for their "Standard Code Signing" product clearly mentions Authenticode compliance:

&nbsp; nbsp; https://www.certum.eu/certum/cert,offer_en_standard_code_signing.xml

If their Open Source one was, it be mentioned. :smile:

**Comment by justinclift** [comment]
I've just emailed them directly to ask.  It could be the case that their page just doesn't list it due to an oversight or something. :smile:

**Comment by NightmareJoker2** [comment]
There are no special **Authenticode™** certificates. It is merely a name for the system Windows uses to verify code signatures. If a certificate is signed by or valid descendant signers of any trusted root authority certificate in the trust store, and bears the *code signing* flag, it can be used to sign code, with, or without a timestamp (which requires a timestamp provider). This includes device drivers (not that this is required here). You can even add your own certification authority and it will work.

You can read more about this here:
1. [Get a code signing certificate](https://docs.microsoft.com/windows-hardware/drivers/dashboard/get-a-code-signing-certificate)
2. [Signing and Checking Code with Authenticode](https://msdn.microsoft.com/library/ms537364(v=vs.85).aspx)
3. [Everything you need to know about Authenticode Code Signing](https://blogs.msdn.microsoft.com/ieinternals/2011/03/22/everything-you-need-to-know-about-authenticode-code-signing/)

Advertising Authenticode™ support typically means whether a certification authority provides a timestamp provider (which is very misleading, to be sure), though you can use as many of those as you want and pick from any of them when you do. Signing with a timestamp is recommended, so the signature can remain valid, even after the signing certificate has expired.

**Comment by justinclift** [comment]
Thanks @NightmareJoker2, those URLs look to be pretty much exactly what we'll need.  I remember having a bad time trying to figure out how to get things done last time around, and those look to be the exactly correct bits. :smile:

With the Certum certificates, they've also responded saying the Open Source ones are Authenticode compatible anyway.  So it indeed looks like Certum will be the ones chosen. :smile:

**Comment by justinclift** [comment]
@NightmareJoker2 Looking at this again, as the current certificate expires in a few months + we're getting near to our next release.

With the Centum one's, I'm still finding it a bit confusing.  Looking at their comparison chart, "Code and application signing for Windows 32 and 64 systems" seems like it's only part of their EV Code Signing option:

![codesigning-comparision](https://user-images.githubusercontent.com/406299/41280312-8e80e28a-6e26-11e8-992b-0e30d56fa76e.png)

Even the "Eliminates system messages like 'Unknown publisher' and 'Unknown software'" doesn't seem to be something their Open Source Code signing cert does.

Very much confused now. :angel:

**Comment by justinclift** [comment]
Pricing wise, the options seem to look like this:

Certum: https://en.sklep.certum.pl/data-safety/code-signing-certificates.html
  * The [MS page](https://docs.microsoft.com/en-us/windows-hardware/drivers/dashboard/get-a-code-signing-certificate) makes it look like the Certum ones are only useful for HW developers.  It's very unclear. :frowning_face: 
  * When looking at pricing, make sure to change the slider to "I do not have reader and cryptographic card", so they include one in the price.
  * "Open Source" code signing (may be useless?): €86.00
  * Standard code signing cert: €169.00
  * EV code signing cert: €389.00

Comodo: https://www.instantssl.com/ssl-certificate-products/code-signing-index.html
  * Comodo code signing cert: £166.95
  * EV code signing cert: £299.00


DigiCert: https://www.digicert.com/order/order-1.php
  * Standard code signing cert: $111.00 USD
  * EV code signing cert: $224.00 USD


Entrust: https://www.entrustdatacard.com/products/digital-signing-certificates/code-signing-certificates
  * "OV" (standard?) code signing cert: $299 USD
  * EV code signing cert: $399 USD


GlobalSign: https://www.globalsign.com/en/code-signing-certificate/
  * Standard code signing cert: $289 USD
  * EV code signing cert: $410 USD


Symantec
  * Doesn't give pricing without providing identifying details.  They have an extremely bad reputation anyway, so they're not really a consideration.

**Comment by assarbad** [comment]
Oh, one more thing. _I think_ the **Code and application signing for Windows 32 and 64 systems** refers to Windows Store applications. Possibly in particular to the app packages (`.appx` etc) rather than the binaries.

**Comment by justinclift** [comment]
Thanks @assarbad, that does help. :smile:

**Comment by tlhackque** [comment]
Welcome to the exciting world of X.509.  Perhaps a few additional comments will help.

The sale and subsequent death of Startcom is unfortunate.  Their business model (pay for validation, not for certificates) was ideal for individuals & open source projects.

The requirement for a hardware token is not (directly) due to certum - it's due to the requirements established by the CA Security Council in [this document](https://casecurity.org/wp-content/uploads/2016/09/Minimum-requirements-for-the-Issuance-and-Management-of-code-signing.pdf)

See section 16.3, which requires that a code signing certificate's private key be stored on a hardware TPM.  It doesn't have to be certum's - though certum may have specific requirements.  

Symantec/DigiCert discontinued issuance of code signing certs for individuals.  I assume the hassle wasn't worth it - as for a while, the malware crowd would buy certificates so their "product" would be easier to install.

Both Startcom and Symantic got in trouble for not following the CA rules on certificate issuance - the browser vendors essentially put them out of business by removing their root certificates from the official trust stores.

The hassle/expense factor is now high enough that I use a certificate from my own private CA.  You can probably get away with that for your nightly builds (a simple self-signed is easier).  Probably not for your releases.

EV is window dressing - expensive.  And can NOT be issued to individuals - you have to be a government or government-recognized business entity.  See [this link](https://cabforum.org/overview-of-the-extended-validation-ssl-vetting-process/)  The requirements are generally more stringent than Organization Validation (OV).  Besides a different root CA, a CA certificate has a OID embedded in it that identifies it as (claiming to be) EV.

Organization validation requires an organization - many open source projects have a registered non-profit entity that makes that doable.  Projects that don't (e.g. an informal group of developers) fall into the cracks between "individual" (and can't get cert) and "business" (costs and hassle of forming one).  Some CAs might accept a DBA (Doing business as) certificate from your local city/town - they tend to be cheap, and don't require an attorney or lawyer.  But they may not be accepted by all CAs (Startcom didn't accept them even for OV).

Informally, certum seems to be the choice of small open source projects.  As noted above, they seem to be willing to validate individuals if there's enough evidence on the web that they are affiliated with a project.  I have no personal experience with them, however.

**Comment by assarbad** [comment]
@tlhackque

> It doesn't have to be certum's - though certum may have specific requirements.

Certum _does_ have special requirements. And for them it has to be their crypto module. TPM isn't a hard requirement, but it's one alternative (to such an external crypto module) option in general. Never used it and I don't think Certum has a way to handle that, at least from what I deduce from how my CSR was handled and all.

> Symantec/DigiCert discontinued issuance of code signing certs for individuals. I assume the hassle wasn't worth it [...]

Uhm, I got one from them a few years ago. As an individual. The hassle was entirely with the individual trying to get the certificate, not so much Symantec :wink: 

It's not as if they every provided any kind of insurance; just the assurance that the entity to which the certificate was issued is who it claims to be. Besides, the requirements were somewhat vague and _very_ US-centric. A notary in the US is something entirely different than in Germany, for example. Oh and of course all documentation needs to be provided in English. Scanned copies, though. How well-trained _are you_ to figure out a faked passport or other document, based on a _scanned copy_ :grin:. I'm sure someone unscrupulous enough to write malware will happily provide scanned fake documents. How would they find out?

Given the hassle it took to get their certificate, I really am happy how straightforward things were with Certum.

> But they may not be accepted by all CAs (Startcom didn't accept them even for OV).

AFAIK not only Certum but also other CAs (I know it for a fact from Comodo) offer the cheaper Open Source code-signing certificates for individuals. Besides, _older_ StartCom certificates had another really subtle and yet really bad problem: they had the `Lifetime Signing (1.3.6.1.4.1.311.10.3.13)` EKU set and therefore signatures became invalid along with the certificate's expiry. Typically time-stamping can be used to counter this (provided that EKU isn't set). Newer certificates didn't have the issue. But it was very annoying and also something you'd typically find out in hindsight.

**Comment by NightmareJoker2** [comment]
@assarbad

Sorry, I'm going to have to chime in here again...

> Downside is that for the first time you need to purchase a USB-token which contains a smartcard which contains the key used for performing the CSR (certificate signing request), which you do after you have purchased the token. The certificate itself costs approx. 30.00 €. The reader and delivery is the expensive part the first time around.

> You need that reader and card *from them*

This is actually wrong.

> The requirement for a hardware token is not (directly) due to certum - it's due to the requirements established by the CA Security Council in [this document](https://casecurity.org/wp-content/uploads/2016/09/Minimum-requirements-for-the-Issuance-and-Management-of-code-signing.pdf)
> 
> See section 16.3, which requires that a code signing certificate's private key be stored on a hardware TPM. It doesn't have to be certum's 

This is partially correct. A SmartCard or other type of hardware security module is also allowed.

> > though certum may have specific requirements.
>
> Certum does have special requirements. And for them it has to be their crypto module.

They do not, with the sole exception, that a new key must be generated using the security token's key generator. This has annoyed me quite a bit, because I like for my public key to remain the same recognizable key and the CA to act as a "trusted introducer" crutch, you know, like they're supposed to and only change it and revoke and replace it when I can say it has been compromised. Also, I like having a backup on another smart card or token. I almost had a fight with their support staff about a year ago over it and ended up writing a smartcard emulation driver just to get around the restriction in the issuance process. Anyway... **Not the point**.

The key generation and certificate issuance process works via the Credential Security Support Provider (CredSSP) Protocol process. This means any PC/SC compatible smart card reader and smart card for which an appropriate reader and smart card driver is installed on the system will work.
They have assured me that they officially support the *Szafir*, *CenCert*, *EuroCert*, and *Sigillum* smart cards, but none of their options offer 4096-bit RSA key support. However, the [openPGP smart card](https://www.floss-shop.de/en/security-privacy/smartcards/13/openpgp-smart-card-v3.3) does support 4096-bit RSA keys, and also, allows for up to three keys per card at once. And best of all, they can be had for ~20 bucks a pop and be used in any PC/SC compatible smart card reader (which are also very inexpensive). Using an openPGP smart card works very well with all certifying authorities I have tested the issuance process with, and the signing tools in Windows, provided either the [openSC](https://github.com/OpenSC/OpenSC) or [openPGP-CSP](https://github.com/vletoux/OpenPGP-CSP) driver are installed. The only problem, the credential front-end in Windows, even in the most recent version, though both driver and credential security provider support exist, does not support passing PIN input to a PIN pad smart card reader and makes use of the native user account control and/or credential prompt instead (complete with press `Ctrl`+`Alt`+`Del` and other types of GPO configuration for it).

> With the Centum one's, I'm still finding it a bit confusing. Looking at their comparison chart, "Code and application signing for Windows 32 and 64 systems" seems like it's only part of their EV Code Signing option:
>
> ![codesigning-comparision](https://user-images.githubusercontent.com/406299/41280312-8e80e28a-6e26-11e8-992b-0e30d56fa76e.png)
>
> Even the "Eliminates system messages like 'Unknown publisher' and 'Unknown software'" doesn't seem to be something their Open Source Code signing cert does.

This is definitely nonsense, and clearly meant for one thing and one thing only: to confuse and make companies with deep pockets spring for the most expensive option at all times, just because it ticks all the boxes.
*Unknown Publisher* in the UAC prompt, Windows Defender's *unknown software* protection and the SmartScreen Filter warnings on downloaded files in Internet Explorer and Microsoft Edge will definitely be eliminated with any code signing certificate, even one that chains up to a root CA you created yourself and have installed into your trust store on your own computer. This is mostly because if you were to sign harmful software, you can be easily held accountable and/or put in jail for doing so, because of the identity verification process tied to the certificate issuance process. **All you should look at in that chart are the second and 4th options, all the others result from those, or are the same thing.** (This excludes cross-signing from Microsoft, which is listed in the EV option, but not actually up to the CA, but Microsoft instead, and required for some types of hardware or kernel mode drivers, and Microsoft will happily cross-sign a non-EV signature and/or provide their own *enhanced validation* for the purpose. You will typically not get the ability to sign arbitrary binaries that require that type of signature yourself and will need to go through WHQL/Windows Logo certification for those or use Authenticode and your own CA for internal purposes instead.)

> *I think* the **Code and application signing for Windows 32 and 64 systems** refers to Windows Store applications. Possibly in particular to the app packages (`.appx` etc) rather than the binaries.

No. Windows Store apps (Universal Windows Platform) are signed by the Windows Store Third-party Signing Authority after review of the submitted application. A developer's signature is not required. If the application is side-loaded, those [application packages](https://docs.microsoft.com/en-us/windows/uwp/packaging/packaging-uwp-apps) can [have a signature](https://docs.microsoft.com/en-us/windows/uwp/packaging/create-certificate-package-signing), but typically, [line of business distribution](https://docs.microsoft.com/en-us/windows/uwp/publish/distribute-lob-apps-to-enterprises) or a [private store](https://docs.microsoft.com/en-us/microsoft-store/distribute-apps-from-your-private-store) within the enterprise is used instead.

**Comment by justinclift** [comment]
> Unknown Publisher in the UAC prompt, Windows Defender's unknown software protection and the SmartScreen Filter warnings on downloaded files in Internet Explorer and Microsoft Edge will definitely be eliminated with any code signing certificate, even one that chains up to a root CA you created yourself and have installed into your trust store on your own computer.

This sounds a bit weird to me.  I've personally downloaded our DB4S 3.10.1 (release) on Win 7 x64 just a few months ago, and the warning thing was still shown in the IE window (before installing  Firefox instead).  DB4S 3.10.1 was definitely signed (with StartCom cert), and this was prior to the cert expiring.

It didn't bother me too much personally, but I did make a mental note of the warning. :wink:

**Comment by NightmareJoker2** [comment]
> and the warning thing was still shown in the IE window

Do you mean the *this software was downloaded from the internet* warning? That's something else, and you can't get rid of that:
![IAttachmentExecute prompt](https://www.tenforums.com/attachments/tutorials/136808d1495820363-disable-downloaded-files-being-blocked-windows-open_file-security_warning.png)

By signing your software, instead of **Unknown Publisher** it will show the name of the signer instead. If the executable is signed, opening the file will no longer be blocked as the default action when SmartScreen is enabled, and on computers where the group policies to block unsigned software are enabled, your software will run or ask for elevation. Users won't see this:
![IAttachmentExecute SmartScreen](https://www.tenforums.com/attachments/tutorials/136806d1495820363-disable-downloaded-files-being-blocked-windows-windows-smartscreen.png)

**Comment by justinclift** [comment]
Heh Heh Heh.  Yeah, that was the one.

Sounds like there are no actual good options, only shades of "barely working" through to "non-functional".  In that case, lets go with the cheapest option to allow people to download it, as all of the other options are crap.

**Comment by justinclift** [comment]
Update! :smile:

We've received the Certcom certificate, and a Certcom smart card + reader to go with it.

Weirdly, it seems to require a Windows PC + Java (wtf kind of "secure" combo is this starting with!?!?!?!?!?!?!?!) .

Anyone have recommendations on a Linux alternative way to work with it, so we can get stuff done in a reasonable OS? :smile:

(CC @MKleusberg, who received the actual smart card + cert and is the unfortunate person having to go through this crap! :grinning: )

**Comment by MKleusberg** [comment]
Going to close this issue as we are now able to sign binaries. Generating the key pair was initially fairly painful but the actual signing is pretty straightforward now :smile:

**Comment by fernandreu** [comment]
Hi there, just wondering if you finally found that Linux alternative you were looking for, ideally without requiring to enter a PIN code manually either. I am in a very similar spot for one of my open-source projects and this page is one of the first results in Google :sweat_smile:

Essentially, I just received the open-source code signing kit from Certum with the card reader, but did not expect having to enter the pin code manually every time I want something signed. I was also hoping I could automate the code signing from a CI/CD pipeline somehow, such as by connecting the card reader to a Raspberry Pi agent.

It seems they do have Linux versions of their proCertum CardManager software [here](https://www.certum.pl/pl/wsparcie/cert_oferta_procertum_cardmanager/) (in Polish), but I am not entirely sure if that is all you would need to make it work. I will give it a try though.

Nice project, by the way!

**Comment by MKleusberg** [comment]
Unfortunately we haven't made any progress on this :frowning: If I remember correctly there are several similarly named application from Certum even for Windows and only one of them worked for our use-case, the others failing for no apparent reason. So the Linux version you found definitely is worth a try but you expectations low :wink: If it does work though please let us know. A Linux versions would be really useful :innocent:

**Comment by fernandreu** [comment]
Just in case it helps you guys:

I am having some issues trying to make the Linux version work (as expected). I am also very doubtful that I can find a way of not having to enter a PIN code manually, in any platform. Due to that, I have focused on trying to automate the steps before and after the actual PIN code entering as much as possible for now.

Essentially, I have made a PowerShell script that finds the latest draft release on GitHub, downloads its assets, signs them (at which point the PIN prompts appear), then re-uploads them. You can see it [here](https://github.com/fernandreu/office-ribbonx-editor/blob/master/build/github-release-sign.ps1). The script already does its job, but there is no error checking to ensure the process went well before deleting the original assets yet.

**Comment by fernandreu** [comment]
Alright, I managed to sign my tool in Linux! See below what I did (sorry if it is too long / detailed / with some things too obvious for you guys). I used Ubuntu 19.10 (x64).

### Part 1: Ensure that the card is detected by `pcsc_scan`

`pcsc_scan` is a command from [pcsc-tools](http://ludovic.rousseau.free.fr/softwares/pcsc-tools/). Before compiling this, I had to compile [pcsc-perl](http://ludovic.rousseau.free.fr/softwares/pcsc-perl/) (`pcsc-tools` might compile fine without that, but the `pcsc_scan` would actually not work without it).

You will also need to install `pcscd` and `libpcsclite1` at some point.

My cryptographic card is ACR101I by ACS, so I went to their webpage and downloaded the official drivers [here](https://www.acs.com.hk/en/driver/141/acr101i-simicro-ccid/). There are packages for some Linux distributions but unfortunately not for Ubuntu 19.10 (and `.deb` packages for earlier Ubuntu versions do not seem to work), so I downloaded [the source code](https://www.acs.com.hk/download-driver-unified/11566/ACS-Unified-Driver-Lnx-Mac-117-P.zip) and compiled it myself.

If everything went well, the `pcsc_scan` command should give an output similar to what [the `pcsc-tools` website](http://ludovic.rousseau.free.fr/softwares/pcsc-tools/) shows when the card is inserted.

### Part 2: Install proCertum CardManager

_Note: if you are not planning to manage certificates and you just want to sign tools using already installed certificates, this might be completely optional; for that, I believe you only need the pkcs11 module which you can directly obtain from [here](https://www.certum.pl/pl/data/download.png)._

I believe any of the versions in the [proCertum website](https://www.certum.pl/pl/wsparcie/cert_oferta_procertum_cardmanager/) should work, but I think I went for the one for glibc 2.15.

The main issue I had is: `sudo ./install-file.bin` gives you a yes/no confirmation message (in Polish) before starting the installation... and then it does nothing. However, without `sudo`, it first complains that it needs admin permissions, then it installs it fine. You might need to install Qt and a few other things, but the GUI will eventually show-up when executing `/opt/proCertumCardManager/proCertumCardManager`:

![Screenshot from 2019-12-03 22-27-21](https://user-images.githubusercontent.com/2480661/70096855-f3e6ae00-161f-11ea-8e9a-7be12f340f22.png)

### Part 3: Sign your tool

This part might be a bit too specific for the C# `.exe` / `.msi` files I am interested in signing, but it might help you as well.

The substitute of Windows' `signtool` that I used is `osslsigncode` (it can be installed in Ubuntu via `apt`). Other programs required are `openssl`, `opensc-pkcs11` (not fully sure; I installed that at some point) and `libp11` (compiled from [here](https://github.com/OpenSC/libp11)).

Finally, my sign command looks like this:

```
osslsigncode -pkcs11engine /usr/lib/x86_64-linux-gnu/engines-1.1/pkcs11.so -pkcs11module /opt/proCertumCardManager/sc30pkcs11-2.0.0.39.r2-MS.so -certs my_cert.spc -h sha2 -pass <my-pin-code> -in <unsigned.exe> -out <signed.exe>
```

If you don't specify the `-pass` command, it will ask you for one in the command line (no external dialogues either). If it goes well, you should just see a `Succeeded` message.

---------------

I am also contacting their support team to see if there is any chance to get an `arm64` version of the card manager, or at least the pkcs11 module, so that I can achieve my main goal of running all this from my Rapsberry Pi.

**Comment by justinclift** [comment]
Well done @fernandreu, that should be helpful. :smile:

**Comment by fernandreu** [comment]
FYI, I wanted to automate all the code signing setup process in Ubuntu because I wanted to replicate it in another (virtual) machine. You can see the result [here](https://github.com/fernandreu/office-ribbonx-editor/blob/master/build/sign/install-ubuntu.sh). Installing PowerShell / the Azure Pipelines agent is not needed for signing files, but rather for the specific way in which I am doing it.
